### ISSUE TITLE
[MNG-6087] Version ranges stop build if minimum version is not available

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7655]
**[John Chesshir](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jchesshir)** opened **[MNG-6087](https://issues.apache.org/jira/browse/MNG-6087?redirect=false)** and commented

When I run for example 'mvn package', there's a dependency that cannot be resolved during Maven's pom.xml evaluation faze; This is a mockup of the part of the pom that causes the problem:

```xml
<dependency>
  <groupId>com.project</groupId>
  <artifactId>artifact</artifactId>
  <version>[2.1.3-270,)</version>
</dependency>
```

This results in the following error issued by Maven:
Failed to execute goal on project MyProject: Could not resolve dependencies for project com.myproject:mymodule:jar:1.3.9-240: Could not find artifact com.project:artifact:jar:2.1.3-270 in myRepo (http://myrepo.local:7392/artifactory/repo)

If I browse the Artifactory repository http://myrepo.local:7392/artifactory/repo, there are several versions of com.myproject:mymodule jars present. However, the owners of that artifact have set up their CI to only keep a rolling snapshot, and version 2.1.3-270 is no longer available.

We are using Maven 3.2.5. It seems that Maven should not require any of the named versions in a version range to actually be present before it can see whether any versions exist that do fit in the range. We only want to have to update our minimum version when there is a breaking change that we have to address, not just when the publishers decide to remove the minimum version from availability.

---

**Affects:** 3.2.5

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Maybe I misunderstand your issue or expectations, but you say this artifact should be used if the version is greater or equal `2.1.3-270` with your range definition...So to be honest I don't see the problem. From my point of view it looks like: Works as designed.

On the other hand someone has decided to delete a released version from a repository which is the problem and not the version range (or may be you have to reconsider the definition of your version range)....cause in Maven and in a repository releases have to be immutable...if this is not the case the whole contract will fail...

Maybe you can explain more in detail what you mean by:

> It seems that Maven should not require any of the named versions in a version range to actually be present before it can see whether any versions exist that do fit in the range. We only want to have to update our minimum version when there is a breaking change that we have to address, not just when the publishers decide to remove the minimum version from availability.

At the moment I'm not sure what you mean in detail...

**Comment by jira-importer** [comment]
**[John Chesshir](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jchesshir)** commented

Karl,
First, thank you for looking at this.

> cause in Maven and in a repository releases have to be immutable...if this is not the case the whole contract will fail...

Perhaps that is so technically, but I think I can safely say this is not practical.  Disk space on internal repositories must be closely monitored, and releases that are beyond a support window must be identified and removed.  (Perhaps you could argue that SNAPSHOTS should be used more to prevent this, which would be a valid point, but I'm not sure I can sell that to every employer I know.)

I don't see why the whole contract should be contingent on the MINIMUM version being available when another version exists within the range.  I think the ability to use dependency (or even Parent) version ranges to keep the contract valid would be of great value.  After all, they are used to automatically bring on an upgrade within a specific range.  Why should that fail to work just because a previous version either doesn't or perhaps never did exist?

For example, why can't one use [2.0.5,) when the first version within that range that was EVER released was 2.0.5-56?

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Hi John,

First disk space is cheap...furthermore this is practical cause it's the contract in a maven repository (Releases are immutable).  SNAPSHOT's are something different.

The now given example is something different than the one in the first description and the example you have given `[2.0.5,)` where the first release was made was `2.0.5-56` works. If you don't have a `2.0.5-56` at all in your example the range must fail cause it can't be fulfilled.
Kind regards
Karl Heinz

**Comment by jira-importer** [comment]
**[John Chesshir](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jchesshir)** commented

Karl,

After further looking at our environment, I have determined that this bug does not actually exist in Maven.  It works as I expected, both for the example case I gave in our conversation and for the case described in the Issue's main description.

The bug was an error in our new CI process, which we are still tweaking.  It caused the version range that I thought was in place to be replaced with a static version in our master repository branch, where it should never have been.  So Maven acted like it was referencing a static version that had been removed from the repository ... because it was.  I replaced the static value with the old range using a specific version that I knew had been removed, and that fixed the problem.

I apologize for the rabbit trail, and I thank you for your time.

Sincerely,
John Chesshir
